MAINT: Update `array_api_compat` submodule for the 2023.12 spec

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

fixes https://github.com/scipy/scipy/issues/21795

#### What does this implement/fix?
<!--Please explain your changes.-->

Update the array-api-compat submodule so that we can use the 2023.12 features.

#### Additional information
<!--Any additional information you think is important.-->

It's rather annoying that we need to track the submodule development, including implementation details in `_understored_py_sources.py`. But I guess it's the meson way, so `¯\_(ツ)_/¯`

ev-br committed:
MAINT: _lib: update the array_api_compat submodule to version 1.9.1

ev-br committed:
BUG: _lib: update the install file list for the array-api-compat update

We actually need to list all files to install in a submodule, and
that includes _private python source files!

tylerjereddy commented on scipy/_lib/meson.build:
```diff
@@ -190,6 +190,7 @@ py3.install_sources(
     'array_api_compat/array_api_compat/numpy/__init__.py',
     'array_api_compat/array_api_compat/numpy/_aliases.py',
     'array_api_compat/array_api_compat/numpy/_typing.py',
+    'array_api_compat/array_api_compat/numpy/_info.py',
```
I needed this locally as well (and the CI fail seems to be needing the `torch` one too):

```diff
--- a/scipy/_lib/meson.build
+++ b/scipy/_lib/meson.build
@@ -163,6 +163,7 @@ py3.install_sources(
     'array_api_compat/array_api_compat/cupy/__init__.py',
     'array_api_compat/array_api_compat/cupy/_aliases.py',
     'array_api_compat/array_api_compat/cupy/_typing.py',
+    'array_api_compat/array_api_compat/cupy/_info.py',
     'array_api_compat/array_api_compat/cupy/fft.py',
     'array_api_compat/array_api_compat/cupy/linalg.py',
   ],
```

if I understand your comment above correctly, I believe this is a best practice in both `meson` and CMake for faster build experience. I do know some teams that use glob patterns instead of maintaining lists, but best practice in both those build systems has mostly been to maintain the lists IIRC.

I believe there is some other options via subprojects or whatever each of them calls it when they provide the appropriate build system-native infra themselves.

tylerjereddy replied:
(so if `array_api_compat` maintained its own meson build system/list like above I think there'd be a way to avoid it perhaps, but Ralf would know better than I; might be dumping too much on Aaron for the benefit though)

rgommers replied:
Tyler is right - instant rebuilds and not installing unnecessary files have a price, and that price is "no globbing". See this FAQ entry: https://mesonbuild.com/FAQ.html#why-cant-i-specify-target-files-with-a-wildcard

> (so if `array_api_compat` maintained its own meson build system/list like above I think there'd be a way to avoid it perhaps, but Ralf would know better than I; might be dumping too much on Aaron for the benefit though)

Indeed, it's too much work upstream, and Meson isn't the build system I would choose for a small pure Python project probably. Maintaining the file listing here is easier overall.

ev-br committed:
TST: cluster: add skips for tests which emit DeprecationWarning from MPL

Under MPL 3.9.2 and torch as the backend, MPL emits

    if (not np.isfinite(vmin)) or (not np.isfinite(vmax)):
E   DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)

ev-br commented:
Thanks Tyler! Pushed an update with
- pin array-api-compat at the 1.9.1 release tag
- updated the file listing in meson.build (your point that the explicit listing is preferable is well taken)
- skipped two dengrogram plotting tests under torch, which emit DeprecationWarnings with MPL 3.9.2, pytorch 2.4 and numpy 2.1; The warnings are emiited from deep inside MPL which does something like `np.isfinite(torch_tensor)` and complain about the signature of `__array_wrap__` having been changed in numpy 2.0. So I suppose the best we can do in scipy is to skip the offending tests and revisit the skips at some point.

rgommers commented:
> The warnings are emiited from deep inside MPL which does something like `np.isfinite(torch_tensor)` and complain about the signature of `__array_wrap__` having been changed in numpy 2.0. So I suppose the best we can do in scipy is to skip the offending tests and revisit the skips at some point.

Indeed, this is annoying and needs a fix in PyTorch - I'll put it on the list.

tylerjereddy commented:
I poked around a bit locally on a Linux box with GPU and an ARM mac, and various array API testing incantations with this branch seemed "ok" to me, and I'm happy with the diff now.

I didn't bother checking with `mps` backend on ARM Mac because that still has hard crashes on `main` anyway (I've been reporting quite a few upstream to `torch`, but those are sometimes even OS-level/shipped fixes apparently).

tylerjereddy commented:
thanks Evgeni, Ralf
